File: packages/gamole_ai/overlap.py
"""Overlap detection - ported from packages/ai/src/overlap.ts."""

import logging
from dataclasses import dataclass, field

from .embeddings import embed_batch

logger = logging.getLogger(__name__)


@dataclass
class OverlapMatch:
    linear_id: str
    title: str
    similarity: float


@dataclass
class OverlapResult:
    story_index: int
    story_title: str
    overlaps: list[OverlapMatch] = field(default_factory=list)


async def detect_overlaps(
    stories: list[dict],
    threshold: float = 0.85,
    top_k: int = 3,
    team_id: str | None = None,
) -> list[OverlapResult]:
    """Detect overlapping/duplicate Linear issues for generated stories.

    G10: Simple cosine similarity only — no re-ranking.
    """
    if not stories:
        return []

    texts = [f"{s['title']} {s['description']}" for s in stories]
    embeddings = await embed_batch(texts)

    results: list[OverlapResult] = []

    try:
        from sqlalchemy import select, text

        from gamole_db import LinearIssueCache, get_session

        async for session in get_session():
            for i, story in enumerate(stories):
                embedding = embeddings[i] if i < len(embeddings) else None
                if not embedding:
                    continue

                embedding_literal = str(embedding)
                similarity_sql = f"1.0 - (embedding <=> '{embedding_literal}'::vector)"

                stmt = (
                    select(
                        LinearIssueCache.linear_id,
                        LinearIssueCache.title,
                        text(f"{similarity_sql} as similarity"),
                    )
                    .where(text(f"{similarity_sql} >= {threshold}"))
                    .order_by(text(f"embedding <=> '{embedding_literal}'::vector"))
                    .limit(top_k)
                )

                if team_id:
                    stmt = stmt.where(LinearIssueCache.team_id == team_id)

                result = await session.execute(stmt)
                matches = [
                    OverlapMatch(
                        linear_id=row.linear_id,
                        title=row.title,
                        similarity=row.similarity or 0,
                    )
                    for row in result
                ]

                results.append(OverlapResult(
                    story_index=i,
                    story_title=story.get("title", ""),
                    overlaps=matches,
                ))

    except Exception:
        logger.warning("detect_overlaps: DB unavailable, returning empty overlaps")
        return [
            OverlapResult(story_index=i, story_title=s.get("title", ""), overlaps=[])
            for i, s in enumerate(stories)
        ]

    return results

File: packages/ai/src/overlap.ts
import { cosineDistance, desc, sql, eq, and } from 'drizzle-orm'
import { embedBatch } from './embeddings'

export interface OverlapResult {
  storyIndex: number
  storyTitle: string
  overlaps: Array<{ linearId: string; title: string; similarity: number }>
}

/**
 * Detect overlapping/duplicate Linear issues for a set of generated stories.
 *
 * For each story, embeds `title + " " + description` and queries the
 * `linear_issues_cache` for the top-K most similar existing issues using
 * cosine similarity. Results above `threshold` are considered potential
 * duplicates.
 *
 * Returns an empty overlaps array for each story if the DB is unavailable.
 * G10: Simple cosine similarity only — no re-ranking.
 *
 * @param stories - Array of stories with title and description
 * @param options - Optional configuration
 * @param options.threshold - Similarity threshold (default: 0.85)
 * @param options.topK - Max similar issues per story (default: 3)
 * @param options.teamId - Filter by Linear team ID
 */
export async function detectOverlaps(
  stories: Array<{ title: string; description: string }>,
  options: { threshold?: number; topK?: number; teamId?: string } = {},
): Promise<OverlapResult[]> {
  const { threshold = 0.85, topK = 3, teamId } = options

  if (!stories.length) return []

  // Embed all stories in batch
  const texts = stories.map((s) => `${s.title} ${s.description}`)
  const embeddings = await embedBatch(texts)

  const results: OverlapResult[] = []

  try {
    const { db, linearIssuesCache } = await import('@gamole/db')

    for (let i = 0; i < stories.length; i++) {
      const embedding = embeddings[i]
      if (!embedding) continue

      const similarity = sql<number>`1 - ${cosineDistance(linearIssuesCache.embedding, embedding)}`

      const conditions = teamId
        ? and(eq(linearIssuesCache.teamId, teamId), sql`${similarity} >= ${threshold}`)
        : sql`${similarity} >= ${threshold}`

      const matches = await db
        .select({
          linearId: linearIssuesCache.linearId,
          title: linearIssuesCache.title,
          similarity,
        })
        .from(linearIssuesCache)
        .where(conditions)
        .orderBy(desc(similarity))
        .limit(topK)

      results.push({
        storyIndex: i,
        storyTitle: stories[i]?.title ?? '',
        overlaps: matches.map((m) => ({
          linearId: m.linearId,
          title: m.title,
          similarity: m.similarity ?? 0,
        })),
      })
    }
  } catch {
    // DB unavailable — return empty overlaps for all stories
    return stories.map((s, i) => ({ storyIndex: i, storyTitle: s.title, overlaps: [] }))
  }

  return results
}
